### DISCUSSION TITLE
Sphinx documentation and submodules

### DISCUSSION THREAD

**Discussion by salahaldin** [pyscaffold/pyscaffold#409] [Q&A] [answered]
Hello Guys, 

Environment: 4.0rc1 

I have added a folder under src/<package>/tools , this folder contains project dependencies which developed locally as git submodules. I have changed the sphinx conf.py as follows 
```
   tools_dir = "tools/*"
   ...
    cmd_line_template = (
        "sphinx-apidoc --implicit-namespaces -f -o {outputdir} {moduledir} ..{toolsdir}"
    )
    cmd_line = cmd_line_template.format(outputdir=output_dir, moduledir=module_dir, toolsdir=tools_dir)

```
This did not work. The reason of putting the dependency inside the src package is to make sure that sphinx can find and resolve imports, but now I am facing issues with generating docs for submodules also. Any hint to solve this or best practice?

**Comment by abravalheri** [comment, marked as answer]
Hello @salahaldin, according to `sphinx-apidoc --help`, the arguments of the command are:
```bash
sphinx-apidoc [OPTIONS] -o <OUTPUT_PATH> <MODULE_PATH> [EXCLUDE_PATTERN, ...]
```

So I believe the command you are trying to compose is telling sphinx to not generate documentation for `tools_dir` (in your example `EXCLUDE_PATTERN=..toos/*`), is that right?
If that is the case, the command might not be working  because the path might be slightly imprecise... Have you considered to use the same technique we use in `module_dir` (`tools_dir = os.path.join(__location__, "../src/tools/*")`)?

If you are instead interested in silence some kind of warnings from sphinx about unknown modules, may I ask you how do you deal with the submodules as dependencies for installing your package? I assume that you either add them to PYTHONPATH or install them in a virtual environment before installing your package... Have you tried to use the same technique in the virtual environment in which sphinx run? If you are using the tasks in the generated `tox.ini` file you could try either the `setenv` directive or adding editable installs to `docs/requirements.txt`.

**Reply by salahaldin** [reply]
Thank you , I have added the exclude pattern and this worked fine.  Still I am not able to get sphinx working due to failed to import warnings due to submodules.  I have issues with imports , I will try older version with less options to have more control of the project structure.

**Reply by abravalheri** [reply]
Hi @salahaldin, have you tried to use the same technique you are using to make your local submodules available as import dependencies to your main project inside the virtualenv running sphinx?

**Reply by salahaldin** [reply]
Hello @abravalheri , I am trying to get a simple solution working. Here is a project structure 

```
test_1 % tree -I __pycache__ 
|-- __init__.py
|-- skeleton.py
|-- test.py
`-- test2
    |-- __init__.py
    `-- test3.py
```
in skeleton.py, I have the following imports

```
import test
from test2 import test3
```
and in the main 
```
if __name__ == "__main__":
    # run()
    print(test3.xxx)
    print(test.a())`
```
When running tox -e docs I get the following 
```
WARNING: autodoc: failed to import module 'skeleton' from module 'test_1'; the following exception was raised:
No module named 'test2'
```

**Reply by FlorianWilhelm** [reply]
Good morning @salahaldin, the code you pasted should not work at all except when you run `skeleton.py` directly from the package directory. You are missing the **relative imports**, i.e. it should read:
```
from . import test
from .test2 import test3
```
instead of the imports in your code. 

I think this should also fix your Sphinx problems.

**Reply by salahaldin** [reply]
Thank you @FlorianWilhelm  and @abravalheri , all the issues are solved. I have created a script to install the dependencies and I have added this script to tox.ini such as 

`commands =
    ./deps_setup.sh`

After that everything worked as charm. Using tox was really cool and it helped me to setup development virtual environment.

Great effort and support :-)
